networkx/algorithms/cluster.py

```diff
@@ -47,17 +49,42 @@ def triangles(G, nodes=None):
 
     Notes
     -----
-    When computing triangles for the entire graph each triangle is counted
-    three times, once at each node.  Self loops are ignored.
+    Self loops are ignored.
 
     """
     # If `nodes` represents a single node in the graph, return only its number
     # of triangles.
     if nodes in G:
         return next(_triangles_and_degree_iter(G, nodes))[2] // 2
-    # Otherwise, `nodes` represents an iterable of nodes, so return a
+
+    # if `nodes` is a container of nodes, then return a
     # dictionary mapping node to number of triangles.
-    return {v: t // 2 for v, d, t, _ in _triangles_and_degree_iter(G, nodes)}
+    if nodes:
+        return {v: t // 2 for v, d, t, _ in _triangles_and_degree_iter(G, nodes)}
+
+    # if nodes is None, then compute triangles for the complete graph
+
+    # dict used to avoid visiting the same nodes twice
+    # this allows calculating/counting each triangle only once
+    later_neighbors = {}
+
+    # iterate over the nodes in a graph
+    for node, neighbors in G.adjacency():
+        later_neighbors[node] = {
+            n for n in neighbors if n not in later_neighbors and n is not node
+        }
+
+    # instantiate with a zero count for each node
+    # add 1 to the count if a nodes neighbor's neighbor is also a neighbor
+    triangle_counts = dict.fromkeys(G, 0)
+    for node1, neighbors in later_neighbors.items():
+        for node2 in neighbors:
+            for node3 in neighbors & later_neighbors[node2]:
+                triangle_counts[node1] += 1
+                triangle_counts[node2] += 1
+                triangle_counts[node3] += 1
```

Comment:
This can be sped up by moving two increments out of the inner loop and using a `collections.Counter` object to push the third increment into compiled `Counter.update` code.
```suggestion
    triangle_counts = collections.Counter()
    for node1, neighbors in later_neighbors.items():
        for node2 in neighbors:
            third_nodes = neighbors & later_neighbors[node2]
            m = len(third_nodes)
            triangle_counts[node1] += m
            triangle_counts[node2] += m
            triangle_counts.update(third_nodes)
```

Reply:
In github you can Edit a message you have written by clicking the three dots next to the emoji smiley face in the upper right. :)

Also, Nice Catch on starting the Counter with a dict keyed by node to the value 0! :} Otherwise we would lose the nodes without any triangles.

And another style issue we often wrestle with is whether to spell out the full word `neighbors` or abbreviate it `nbrs`.  The long way is "better" because nobody has to guess what it is. The short way helps keep the code from being split over many lines. We often use `nbrs` when it helps reduce line-splitting. And we often use `neighbors` when it doesn't.  We can get away with that because `nbrs` is a pretty unmistakable abbreviation. 
:)